File: leetcode/python/347_top_k_frequent_elements.py
"""
LeetCode Problem : Top K Frequent Elements
Topic            : Array, Hash Table, Divide and Conquer, Sorting, Heap (Priority Queue), Bucket Sort, Counting, Quickselect
Level            : Medium
URL              : https://leetcode.com/problems/top-k-frequent-elements/
Description      : Given an integer array nums and an integer k, return the k most frequent elements. You may return
        the answer in any order.
Examples         :
        Example 1:
        Input: nums = [1,1,1,2,2,3], k = 2
        Output: [1,2]

        Example 2:
        Input: nums = [1], k = 1
        Output: [1]
"""

from collections import Counter
from typing import List


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        """Finds the k most frequent elements in the given list of integers using bucket sort.

        Args:
            nums (List[int]): A list of integers.
            k (int): The number of most frequent elements to return.

        Returns:
            List[int]: A list containing the k most frequent elements in descending order of frequency.

        Time Complexity:
            O(n): Where n is the length of the input list, as we iterate through the array once.

        Space Complexity:
            O(n): Space used for the hash map and frequency buckets.
        """

        count = Counter(nums)
        max_freq = max(count.values()) if count else 0
        buckets = [[] for _ in range(max_freq + 1)]

        for num, freq in count.items():
            buckets[freq].append(num)

        result = []
        for freq in range(max_freq, 0, -1):
            if buckets[freq]:
                result.extend(buckets[freq])
                if len(result) >= k:
                    break

        return result[:k]


def run_tests():
    inputs = {"case_1": [[1, 1, 1, 2, 2, 3], 2], "case_2": [[1], 1]}
    outputs = {"case_1": [1, 2], "case_2": [1]}

    solution = Solution()

    for case, input in inputs.items():
        result = solution.topKFrequent(input[0], input[1])
        assert result == outputs[case], f"{case}: expected {outputs[case]}, got {result}"

    print("✅ All tests passed!")


if __name__ == "__main__":
    run_tests()

File: leetcode/go/347_top_k_frequent_elements.go
package main

import (
	"fmt"
	"os"
)

func topKFrequent(nums []int, k int) []int {
	frequent := make(map[int]int, len(nums))
	for _, num := range nums {
		frequent[num]++
	}

	maxFrequent := 0
	bucket := make([][]int, len(nums)+1)
	for num, freq := range frequent {
		maxFrequent = max(maxFrequent, freq)
		bucket[freq] = append(bucket[freq], num)
	}

	result := make([]int, 0)
	for i := maxFrequent; i >= 0; i-- {
		if len(bucket[i]) == 0 {
			continue
		}

		result = append(result, bucket[i]...)
		if len(result) >= k {
			break
		}
	}

	return result[:k]
}

func RunTestTopKFrequentElements() {
	testCases := map[string]struct {
		nums   []int
		k      int
		expect []int
	}{
		"case-1": {
			nums:   []int{1, 1, 1, 2, 2, 3},
			k:      2,
			expect: []int{1, 2},
		},
		"case-2": {
			nums:   []int{1},
			k:      1,
			expect: []int{1},
		},
		"case-3": {
			nums:   []int{1, 2, 1, 2, 1, 2, 3, 1, 3, 2},
			k:      2,
			expect: []int{1, 2},
		},
	}

	for name, testCase := range testCases {
		fmt.Printf("RUN %s\n", name)

		result := topKFrequent(testCase.nums, testCase.k)
		if !EqualSlices(result, testCase.expect) {
			fmt.Printf("=== FAILED: expect = %v - got = %v\n", testCase.expect, result)
			os.Exit(1)
		}
		fmt.Printf("=== PASSED\n")
	}

	fmt.Printf("\n✅ All tests passed!\n")
}
